ceph/rados/serviceability_workflows.py

```diff
@@ -386,3 +280,68 @@ def remove_osds_from_host(self, host_obj):
             log.error(f"Failed with exception: {e.__doc__}")
             log.exception(e)
             raise
+
+    def drain_host(self, host_obj, zap=True, keep_conf=False):
+        """
+        Method to execution drain on a particular host
+        Args:
+            host_obj: node object of the cluster host
+            zap: flag to control --zap-osd-devices
+            keep_conf: flag to control --keep-conf-keyring
+        Returns:
+            True -> drain successful | Fail -> drain failed
+        """
+        # Starting to drain a host.
+        log.info("Starting to drain host: ", host_obj.hostname)
+        cmd = "ceph orch host drain %s --force" % host_obj.hostname
+
+        if zap:
+            cmd = cmd + "--zap-osd-devices"
+        if keep_conf:
+            cmd = cmd + "--keep-conf-keyring"
+
+        # enable faster recovery
+        self.rados_obj.change_recovery_threads(config={}, action="set")
+
+        self.cephadm.shell(args=[cmd])
+        # Sleeping for 2 seconds for removal to have started
+        time.sleep(2)
+        log.debug("Started drain operation on node : ", host_obj.hostname)
+
+        status_cmd = "ceph orch osd rm status"
+        end_time = datetime.datetime.now() + datetime.timedelta(seconds=3600)
+        drain_success = True
+        while end_time > datetime.datetime.now():
+            try:
+                drain_ops = self.rados_obj.run_ceph_command(
+                    cmd=status_cmd, client_exec=True
+                )
+                for entry in drain_ops:
+                    if entry["drain_done_at"] is None or entry["draining"]:
+                        log.info(
+                            "Drain operations are going on host %s \nOperations: %s"
+                            % (host_obj.hostname, entry)
+                        )
+                        log.debug(
+                            "drain process for OSD %s is still going on"
+                            % entry["osd_id"]
+                        )
+                        if end_time < datetime.datetime.now():
```

Comment:
Can we have a specific end time for each OSD? like maybe 15 - 20 mins, instead of using the endtime defined for the entire drain operation.

Reply:
When drain is triggered, all the OSDs are drained at the same time, if the drain process was sequential or if each OSD was drained individually we could have waited for each OSD separately. Although we are checking the drain process for each daemon, the drain itself is a singular process, therefore having a global timeout helps

Let's suppose, OSD 5 draining takes 30 mins, in these 30 mins, other OSDs and daemons will also be drained. But If we add timeout for individual OSD, it might start failing. With current implementation, we are providing each daemon the complete 1 hour timeout for its drain to complete

Please let me know your thoughts

Reply:
Thanks for the explanation. Current implementation should be fine.